Complete Allblocks

dlesnoff opened the pull request:
Fix #194.
The `nbFile` and `nbShow` blocks are missing in the documentation but could be added after the PR.
I am not sure of their usage and was not very creative for examples today.

I added a TOC to list all blocks at the top and give quick access to their documentation.

The whale image (chosen for some kind of consistency with the icon) is licensed under [the Unsplash license](https://unsplash.com/license).

The documentation now requires Nimpy to work. which makes nimpy a *soft* dependency.

I removed the blocks description from the README.md.
The documentation link won't work until the docs are added to pietroppeter's blog.

Other partials (blocks too?) in `renders.nim` to add in `allblocks.nim`:
- [x] nbText
- [x] nbCode
- [x] nbFile
- [ ] nbJsFromCode
- [x] nimibCode

HugoGranstrom commented on docsrc/allblocks.nim:
```diff
@@ -1,40 +1,126 @@
+
+nbCodeBlock:"nbTextWithCode"
+nbText:"""
+You can run Nim code directly in your markdown text with this block. It may be of use for instance for computation inside tables. See `numerical.nim`.
```
This paragraph is incorrect. What you describe is string interpolation using the `fmt` macro which works with ordinary `nbText` equally well. `nbTextWithCode` differs from `nbText` only by the fact that the block also stores the code that produced the string passed to it. So visually and functionally there is no difference between `nbText` and `nbTextWithCode` for the average user.

dlesnoff replied:
When is the code stored, reused and not reused? When you recompile your code, does it know what it can reuse code?
If there is no difference for the average user, what is the point of this block specifically?
How would you explain this block then? May I ask you to write a suggested recommendation, so I can commit it directly with Github?

HugoGranstrom replied:
> When is the code stored, reused and not reused? When you recompile your code, does it know what it can reuse code? If there is no difference for the average user, what is the point of this block specifically?

Everything in nimib is a block, a call to `nbText` creates a new block containing the text and a call to `nbCode` creates a block containing the code and captured output. But what if you want to get the code passed to `nbText` as well and extract it? `nbText` doesn't store the code in its block (it did in the past but not anymore) so we created `nbTextWithCode` as an alternative when you want to not only get the string but also the code that created the string. The code `fmt"{1+1}"` produces the string `"2"`. `nbText` only gives us the `"2"` but `nbTextWithCode` stores the `fmt"{1+1}"` as well. The code can be accessed with `nb.blk.code` right after creating the `nbTextWithCode`. 

> How would you explain this block then? May I ask you to write a suggested recommendation, so I can commit it directly with Github?

Good question, I'll have to think about it. I'll make a suggestion tonight or tomorrow.

HugoGranstrom replied:
```suggestion
`nbText` only stores the string it is given, but it doesn't store the code passed to `nbText`. For example, `nbText: fmt"{1+1}"` only stores the string `"2"` but not the code `fmt"{1+1}"` that produced that string. `nbTextWithCode` works like `nbText` but it also stores the code in the created block. It can be accessed with `nb.blk.code` right after the `nbTextWithCode` call. See the end of `numerical.nim` for an example.
```

Something like this, perhaps?

dlesnoff replied:
I commit it but I will move it at the bottom of the block's list.
It is more subtle than other blocks and will make a good transition to a "how to create a new block section".
I might detail one example of `nb.blk.code` for the explanation to be complete after.
Thank you for the explanation.

HugoGranstrom commented on docsrc/allblocks.nim:
```diff
@@ -1,40 +1,126 @@
+
+nbRawHtml:"""<h2 style="text-align: center">Centered title</h2>"""
+
+nbCodeBlock:"nbPython"
```
We are planning to move `nbPython` to nimibex (soon hopefully), so I don't know if we should document it in nimib. @pietroppeter any opinions?

pietroppeter replied:
I would skip it here

dlesnoff replied:
Removed.

HugoGranstrom commented on docsrc/allblocks.nim:
```diff
@@ -1,40 +1,126 @@
+nbCodeBlock:"nbCodeSkip"
+nbText: """
 Similar to `nbCode`, `nbCodeSkip` is a block that displays 
 highlighted code but does not compile or run it.
```
Since we (me and Pietro) use the fact that nimib can garantuee that your code compiles as one of the big arguments for using nimib, I think it would be worth it to add a big disclaimer on `nbCodeSkip` that makes that **very** clear. Something like:
```md
> `nbCodeSkip` does **not** garantuee in any way that the code that you show will compile nor run.
```

pietroppeter replied:
I am not sure I feel **that** strongly about this 😉, but more emphasis would not hurt :)

dlesnoff replied:
I used a ⚠️ emoji for this.

HugoGranstrom reviewed (changes requested):
All in all a good summary of the blocks, thank you. Have have written a few comments on adding, changing and perhaps removing some parts though.

HugoGranstrom commented:
>  nbText
> nbCode
> nbFile
> nbShow
> nbJsFromCode

All of these blocks should preferably be added, but that could be done in a follow-up PR by someone else as well :smile:

pietroppeter commented:
> All of these blocks should preferably be added

I would skip `nbShow` which is not a NbBlock (it is more a command like nbInit) and should be deprecated (it is equivalent to nbSave with --nbShow runtime option). The others yeah should be added there.

HugoGranstrom commented:
@pietroppeter `nbShow(df)` 😉

pietroppeter commented on README.md:
```diff
@@ -154,19 +154,7 @@ Currently most of the documentation on customization is given by the examples.
-* `nbClearOutput`: clears the output of preceding code block,
-  useful in case a previous command has produced output that you do not want to show for some reason.
-
+You can find a complete description of the code blocks along with examples in [allblocks.nim](https://pietroppeter.github.io/nimib/allblocks.html)
```
as mentioned in the contributing.md, you need to change index.nim to modify the readme

dlesnoff replied:
I changed the index.nim file too.

dlesnoff replied:
I made one extra change in the index.nim file though. I changed the title to
```
### other blocks
````

pietroppeter commented on docsrc/allblocks.nim:
```diff
@@ -1,40 +1,126 @@
-### nbCodeSkip
+addToc()
 
+nbCodeBlock:"nbCodeSkip"
```
```suggestion
nbCodeBlock: "nbCodeSkip"
```
looks a bit weird to have no space between `:` and the string (also in all other blocks below)

pietroppeter commented on docsrc/allblocks.nim:
```diff
@@ -1,40 +1,126 @@
 """
 
-nbCodeSkip: 
+nbCodeSkip:
```
here and for all the blocks we should probably use `nimibCode` to show the nimib code when calling the blocks (and we should document this usage of nimibCode).

dlesnoff replied:
The distinction between the code of the block and the displayed block is confusing. Plus, it makes the code appears twice with not much addition (the title of the block).
A sentence separating both looks better to me. I still used nimibCode in my latest commits (to come) just so that you can compare.

pietroppeter commented on docsrc/allblocks.nim:
```diff
@@ -1,40 +1,126 @@
+
+nbCodeBlock:"nbClearOutput"
+nbText:"""
+Clears the output of the preceding code block, which is useful if you produce too long output.
```
not just too long but also output you do not want to show (and maybe it is generated from a library), the important use case which should probably be cited explicitly is the case of ggplotnim, which writes to output everytime you save an image

dlesnoff replied:
Fixed in a coming commit.

pietroppeter reviewed (changes requested):
thanks for working on this, I left a few comments to be addressed

dlesnoff commented:
I merged `main` into my branch (integrating my 2 last PRs).
I added nbText, nbCode, nbFile, nimibCode blocks.
I replaced every markdown code with the nimibCode block.
I removed references to nbPython code block and the nimpy dependency.
Various fixes: 
 - Centered title example is smaller. It was too distracting.
 - Removed the incomplete document warning.
Please note that the nbFile example generates an extra file not even in the right directory.
It still lacks the nbJSFromCode block but I leave it to you a good JS example to illustrate it.
I did not understand, do we really drop the nbShow block when there is a special usage in one other document ?

I resolved all the conversations, made the requested changes and I am ready for another review.

dlesnoff commented on docsrc/allblocks.nim:
```diff
@@ -1,40 +1,134 @@
-  echo "The code is not compiled or executed so no output is generated!"
+nbCodeSkip:
+  while true:
+    echo "Notice how there's no output?"
```
I have to change the `is` elision

HugoGranstrom reviewed (changes requested):
This looks great! :D I couldn't find many things to note this time around. Just make sure to add `allblocks.nim` to the docs and we will make sure the preview looks ok as well. And run `nimble readme` locally and push the updated `README.md`.

dlesnoff commented:
I made the requested changes ;)

pietroppeter commented:
Thanks looks very good! The link to see the all blocks preview is https://bc467ce645e09423fd70--nimib.netlify.app (the link we have is an absolute url - since it has to work also for readme -  so it does not work).
Very minor thing: you should use nimibCode also for nbCodeSkip

HugoGranstrom commented:
Once you fix Pietro's minor thing I'm content with merging this

dlesnoff commented:
> The link to see the all blocks preview is https://bc467ce645e09423fd70--nimib.netlify.app/ (the link we have is an absolute url - since it has to work also for readme - so it does not work).

Sorry, I am not sure to understand this. Do you want me to change the link to the allblocks.nim document in the index.nim (and README.md) ?
This link does not point to an allblocks.nim but to the README.md of my PR.

dlesnoff commented:
> you should use nimibCode also for nbCodeSkip

I fixed that! I am done with this PR.
To summarize, we should after this:
- [ ] reopen the issue https://github.com/pietroppeter/nimib/issues/194 for the two last blocks nbShow and nbJSFromCode.
- [ ] Add the allblocks document to pietroppeter.github.io website?
- [ ] Maybe make nimibCode separation between nimibCode rendered code block, the highlighted code and output of these code blocks more explicit? It would be cool to have the name of the files appearing in the code blocks (notebook.nim, notebook.html).

pietroppeter commented:
> > The link to see the all blocks preview is https://bc467ce645e09423fd70--nimib.netlify.app/ (the link we have is an absolute url - since it has to work also for readme - so it does not work).
> 
> 
> 
> Sorry, I am not sure to understand this. Do you want me to change the link to the allblocks.nim document in the index.nim (and README.md) ?
> 
> This link does not point to an allblocks.nim but to the README.md of my PR.

Yeah, I forgot the important detail: https://b41374816c22d6cd819a--nimib.netlify.app/allblocks.html

Once you go to the link to latest netlify preview (and I updated with latest now), you can add allblocks.html to see the preview of that page.

pietroppeter commented:
> > you should use nimibCode also for nbCodeSkip
> 
> 
> 
> I fixed that! I am done with this PR.

Thanks a lot, will merge this now!

> 
> To summarize, we should after this:
> 
> - [ ] reopen the issue https://github.com/pietroppeter/nimib/issues/194 for the two last blocks nbShow and nbJSFromCode.

👍 (nbJsFromCode has multiple variants, they are all already documented in interactivity.nim so not urgent)
> 
> - [ ] Add the allblocks document to pietroppeter.github.io website?

This should already happen automatically once this is merged

> - [ ] Maybe make nimibCode separation between nimibCode rendered code block, the highlighted code and output of these code blocks more explicit? It would be cool to have the name of the files appearing in the code blocks (notebook.nim, notebook.html).

Yeah, we could customize partial of nimibCode for this document in order to have something like:

⬇️⬇️⬇️ source
(Here the source code)
⬇️⬇️⬇️ output
(Here the output)
